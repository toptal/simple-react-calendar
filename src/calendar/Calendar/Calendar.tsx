import React, { FC, Fragment, SyntheticEvent, useCallback, useEffect, useState } from 'react'

import {
  IDate,
  IDateRange,
  IDayOfWeekRenderProps,
  IDayRenderProps,
  IDaysOfWeekRenderProps,
  IMonthHeaderRenderProps,
  IMonthRenderProps,
  INoticeType,
  ISelectionRange,
  IWeekRenderProps,
} from '../../@types'
import { BLOCK_CLASS_NAME, DAYS_OF_WEEK } from './consts'
import {
  getActiveMonth,
  getAsDate,
  getDay,
  getDayEachDay,
  getDayIsBefore,
  getDayIsCurrentMonth,
  getDayIsDisabled,
  getDayIsHighlighted,
  getDayIsNextMonth,
  getDayISOFormatter,
  getDayIsSame,
  getDayIsSelectable,
  getDayIsSelection,
  getDayIsWeekend,
  getHighlighted,
  getIntervalsAsDate,
  getIsRangeOverlap,
  getMonth,
  getMonthFormattedName,
  getMonthWeeks,
  getSelectedAsDate,
  getSelection,
} from './helper'
import Month from './month'
import Notice from './notice'

export type Props = {
  activeMonth: IDate
  blockClassName: string
  daysOfWeek: string[]
  disabledIntervals?: IDateRange[]
  highlightedIntervals?: IDateRange[]
  maxDate: IDate
  minDate: IDate
  minNumberOfWeeks?: number
  mode?: 'range' | 'single'
  onDayEnter?: (day: Date) => void
  onDayDisabledEnter?: (day: Date) => void
  onMonthChange?: (day: Date) => void
  onNoticeChange?: (noticeType?: INoticeType) => void
  onSelect?: (selected: IDate | ISelectionRange) => void
  onSelectionProgress?: (selected: IDate | ISelectionRange) => void
  selectionRangeLimit?: number
  renderDay?: IDayRenderProps
  renderDayOfWeek?: IDayOfWeekRenderProps
  renderDaysOfWeek?: IDaysOfWeekRenderProps
  renderMonth?: IMonthRenderProps
  renderMonthHeader?: IMonthHeaderRenderProps
  renderWeek?: IWeekRenderProps
  selected?: IDate | ISelectionRange
  showDaysOfWeek: boolean
  today: IDate
  weekStartsOn: 0 | 1 | 2 | 3 | 4 | 5 | 6
}

const Calendar: FC<Props> = (props) => {
  const {
    activeMonth: propActiveMonth,
    blockClassName,
    daysOfWeek,
    disabledIntervals: propDisabledIntervals,
    highlightedIntervals: propHighlightedIntervals,
    maxDate: propMaxDate,
    minDate: propMinDate,
    minNumberOfWeeks,
    mode,
    onDayEnter,
    onDayDisabledEnter,
    onMonthChange,
    onNoticeChange,
    onSelect,
    onSelectionProgress,
    selectionRangeLimit,
    renderDay,
    renderDayOfWeek,
    renderDaysOfWeek,
    renderMonth,
    renderMonthHeader,
    renderWeek,
    selected: propSelected,
    showDaysOfWeek,
    today: propToday,
    weekStartsOn,
  } = props
  /**
   * Convert String and Number to Date instance
   */
  const maxDate = getAsDate(propMaxDate)
  const minDate = getAsDate(propMinDate)
  const activeMonth = getAsDate(propActiveMonth)
  const today = getAsDate(propToday)
  const selected = getSelectedAsDate(propSelected)
  const highlightedIntervals = getIntervalsAsDate(propHighlightedIntervals)
  const disabledIntervals = getIntervalsAsDate(propDisabledIntervals)

  /**
   * Hooks
   */
  const [stateSelection, setStateSelection] = useState<Date | ISelectionRange | undefined>(undefined)
  const [stateNoticeType, setNoticeType] = useState<INoticeType | undefined>(undefined)
  const [stateActiveMonth, setStateActiveMonth] = useState<Date>(
    getActiveMonth({selected, activeMonth, mode: mode as 'range' | 'single', today})
  )

  useEffect(() => {
    setStateActiveMonth(getActiveMonth({activeMonth, selected, mode: mode as 'range' | 'single', today}))
    setStateSelection(getSelectedAsDate(propSelected))
  }, [propActiveMonth, today, propSelected, activeMonth, selected, mode])

  const {start, end} = getHighlighted({highlightedIntervals})

  /**
   * Stored Helper Logic Functions
   */
  const useGetSelection = useCallback(
    () => getSelection({selected, onSelectionProgress, stateSelection, mode: mode as 'range' | 'single'}),
    [selected, onSelectionProgress, stateSelection, mode]
  )
  const useGetDayIsHighlighted = useCallback(
    (date: Date) => getDayIsHighlighted({date, startDay: start, endDay: end}),
    [end, start]
  )
  const useGetDayIsSelectable = useCallback(
    (date: Date) => getDayIsSelectable({maxDate, minDate, date, disabledIntervals}),
    [maxDate, disabledIntervals, minDate]
  )
  const useGetDayIsCurrentMonth = useCallback((date: Date) => getDayIsCurrentMonth({date, activeMonth}), [activeMonth])
  const useGetDayIsNextMonth = useCallback((date: Date) => getDayIsNextMonth({date, activeMonth}), [activeMonth])
  const useGetDayIsDisabled = useCallback((date: Date) => getDayIsDisabled({date, disabledIntervals}), [
    disabledIntervals,
  ])
  const useGetDayIsSelected = useCallback(
    (day: Date) =>
      getDayIsSelection({
        day,
        selection: getSelection({selected, onSelectionProgress, stateSelection, mode: mode as 'range' | 'single'}),
      }),
    [mode, onSelectionProgress, selected, stateSelection]
  )
  const useGetDayIsSelectedEnd = useCallback(
    (day: Date) =>
      getDayIsSelection({
        day,
        type: 'end',
        selection: getSelection({selected, onSelectionProgress, stateSelection, mode: mode as 'range' | 'single'}),
      }),
    [mode, onSelectionProgress, selected, stateSelection]
  )
  const useGetDayIsSelectedStart = useCallback(
    (day: Date) =>
      getDayIsSelection({
        day,
        type: 'start',
        selection: getSelection({selected, onSelectionProgress, stateSelection, mode: mode as 'range' | 'single'}),
      }),
    [mode, onSelectionProgress, selected, stateSelection]
  )
  const useGetDayIsToday = useCallback((day: Date) => getDayIsSame({date: today as Date, dateToCompare: day}), [today])
  const useGetDayIsWeekend = useCallback((day: Date) => getDayIsWeekend({day}), [])
  const useGetWeekMonthWeeks = useCallback(
    (activeMonth: Date) => getMonthWeeks({activeMonth, minNumberOfWeeks, weekStartsOn}),
    [minNumberOfWeeks, weekStartsOn]
  )
  const useGetMonthFormattedName = useCallback((day: Date) => getMonthFormattedName(day), [])
  const useGetDayISOFormatter = useCallback((day: Date) => getDayISOFormatter(day), [])
  const useGetDayEachDay = useCallback(({start, end}: IDateRange) => getDayEachDay({start, end}), [])

  /**
   * Event Handlers
   */
  const handleOnSwitchMonth = ({
    currentTarget: {
      dataset: {calendarValue},
    },
  }: SyntheticEvent<HTMLElement>) => {
    const date = getMonth({date: stateActiveMonth, offsetInMonth: Number(calendarValue)})

    onMonthChange && onMonthChange(date)
    setStateActiveMonth(date)
  }

  // const handleOnNoticeChange = (noticeType?: INoticeType) => {
  //   onNoticeChange && onNoticeChange(noticeType)
  //   setNoticeType(noticeType)
  // }

  const handleOnSelectionChange = (event: SyntheticEvent<HTMLButtonElement>) => {
    console.log('>>>> handleOnSelectionChange', event)
    // const {start, end, inProgress} = changes

    // if (onSelect && start && (mode !== 'range' || !inProgress)) {
    //   onSelect(mode === 'single' ? start : {start, end})
    // }

    // if (mode === 'range') {
    //   if (onSelectionProgress) {
    //     onSelectionProgress(changes)
    //   } else {
    //     setStateSelection(inProgress ? {start, end} : undefined)
    //   }
    // }
  }

  const handleOnDayClick = ({
    currentTarget: {
      dataset: {calendarValue},
    },
  }: SyntheticEvent<HTMLButtonElement>) => {
    const day = getAsDate(calendarValue as string)

    onSelect && onSelect(day)

    if (mode === 'range') {
      const {start, inProgress} = stateSelection as ISelectionRange

      if (inProgress) {
        const isDisabledWithin = getRangeIncludesDisabledInterval({start, end: day, disabledIntervals})

        if (isDisabledWithin) {
          setStateSelection({start: undefined, end: undefined, inProgress: false})
          setNoticeType('overlapping_with_disabled')

          return
        } else {
          setStateSelection({...(stateSelection as ISelectionRange), end: day, inProgress: false})
        }
      } else {
        setStateSelection({start: day, inProgress: true, end: undefined})
      }
    } else {
      setStateSelection(day)
    }

    setNoticeType(undefined)
  }

  const handleOnDayMouseEnter = ({
    currentTarget: {
      dataset: {calendarValue},
    },
  }: SyntheticEvent<HTMLButtonElement>) => {
    const day = getAsDate(calendarValue as string)
    const {inProgress, start, end} = stateSelection as ISelectionRange

    onDayEnter && onDayEnter(day)

    if (!inProgress) return

    const isDisabledWithin = getIsRangeOverlap({selection: stateSelection as ISelectionRange, day})

    if (!isDisabledWithin) return

    const dateLimit = getDay({start, offset: selectionRangeLimit})

    if (getDayIsSame({day1: day, day2: end})) {
      if (!selectionRangeLimit || (selectionRangeLimit && getDayIsBefore({day1: day, date2: dateLimit}))) {
      }
    }
  }

  const handleOnDisabledDayClick = ({
    currentTarget: {
      dataset: {calendarValue},
    },
  }: SyntheticEvent<HTMLButtonElement>) => {
    const day = getAsDate(calendarValue as string)

    onDayDisabledHover && onDayDisabledHover(day)
    console.log('>>>> handleOnDisabledDayClick', event)
  }

  const handleOnDisabledDayHover = ({
    currentTarget: {
      dataset: {calendarValue},
    },
  }: SyntheticEvent<HTMLButtonElement>) => {
    const day = getAsDate(calendarValue as string)

    onDayDisabledHover && onDayDisabledHover(day)
    console.log('>>>> handleOnDisabledDayClick', event)
  }

  const children = (
    <Fragment>
      {!!stateNoticeType && <Notice blockClassName={blockClassName} type={stateNoticeType as INoticeType} />}
      <Month
        activeMonth={stateActiveMonth}
        blockClassName={blockClassName}
        customRender={renderMonth}
        daysOfWeek={daysOfWeek}
        showDaysOfWeek={showDaysOfWeek}
        getDayEachDay={useGetDayEachDay}
        getDayIsCurrentMonth={useGetDayIsCurrentMonth}
        getDayIsDisabled={useGetDayIsDisabled}
        getDayIsHighlighted={useGetDayIsHighlighted}
        getDayIsNextMonth={useGetDayIsNextMonth}
        getDayISOFormatter={useGetDayISOFormatter}
        getDayIsSelectable={useGetDayIsSelectable}
        getDayIsSelected={useGetDayIsSelected}
        getDayIsSelectedEnd={useGetDayIsSelectedEnd}
        getDayIsSelectedStart={useGetDayIsSelectedStart}
        getDayIsToday={useGetDayIsToday}
        getDayIsWeekend={useGetDayIsWeekend}
        getWeekMonthWeeks={useGetWeekMonthWeeks}
        handleOnDayClick={handleOnDayClick}
        handleOnDayMouseEnter={handleOnDayMouseEnter}
        handleOnDisabledDayClick={handleOnDisabledDayClick}
        renderDay={renderDay}
        renderDayOfWeek={renderDayOfWeek}
        renderDaysOfWeek={renderDaysOfWeek}
        renderWeek={renderWeek}
        weekStartsOn={weekStartsOn}
      />
    </Fragment>
  )

  return <div className={blockClassName}>{children}</div>
}

Calendar.defaultProps = {
  blockClassName: BLOCK_CLASS_NAME,
  daysOfWeek: DAYS_OF_WEEK,
  mode: 'single',
  showDaysOfWeek: true,
  today: new Date(),
  weekStartsOn: 1,
}

Calendar.displayName = 'Calendar'

export default Calendar
