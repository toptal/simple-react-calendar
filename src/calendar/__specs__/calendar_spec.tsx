import { shallow } from 'enzyme'
import React from 'react'
import ReactTestRenderer from 'react-test-renderer'

import Calendar from '../calendar'
import Month from '../month'

describe('Calendar', () => {
  let instance, props, wrapper
  const date = new Date(2015, 6, 1)

  describe('#componentWillReceiveProps', () => {
    describe('when nextProps `activeMonth` is defined', () => {
      describe('when nextProps `activeMonth` and prop `activeMonth` is different', () => {
        it('sets state `activeMonth`', () => {
          props = { activeMonth: date }
          wrapper = shallow(<Calendar />)
          wrapper.setProps({ activeMonth: new Date(2015, 9, 1) })

          expect(wrapper.state('activeMonth')).toEqual(new Date(2015, 9, 1))
        })
      })
    })

    describe('when nextProps `activeMonth` is `undefined`', () => {
      it("doesn't set state", () => {
        props = { activeMonth: date }
        wrapper = shallow(<Calendar />)
        wrapper.setProps({ disableDaysOfWeek: false })

        expect(wrapper.state('activeMonth')).not.toEqual(new Date(2015, 9, 1))
      })
    })
  })

  describe('#_initialMonth', () => {
    describe('when prop `activeMonth` is valid date', () => {
      it('returns `activeMonth`', () => {
        wrapper = shallow(<Calendar activeMonth={date} />)

        expect(wrapper.instance()._initialMonth()).toEqual(date)
      })
    })

    describe('when prop `activeMonth` is invalid date', () => {
      it('returns `activeMonth`', () => {
        props = {
          activeMonth: '2017-01-01',
          selected: 'abc',
          today: date
        }
        wrapper = shallow(<Calendar {...props} />)

        expect(wrapper.instance()._initialMonth()).toEqual(props.today)
      })
    })
  })

  describe('#_switchMonth', () => {
    describe('when prop `onMonthChange` is defined', () => {
      it('calls prop #onMonthChange', () => {
        props = { activeMonth: date, onMonthChange: jest.fn() }
        wrapper = shallow(<Calendar {...props} />)
        wrapper.instance()._switchMonth('2017-01-01')

        expect(props.onMonthChange).toHaveBeenCalledTimes(1)

        expect(props.onMonthChange).toHaveBeenCalledWith('2017-01-01')
      })
    })

    describe('when prop `onMonthChange` is `undefined`', () => {
      it('sets `activeMonth` state', () => {
        props = { activeMonth: date }
        wrapper = shallow(<Calendar {...props} />)
        wrapper.instance()._switchMonth('2017-01-01')

        expect(wrapper.state('activeMonth')).toBe('2017-01-01')
      })
    })
  })

  describe('#_activeMonth', () => {
    describe('when prop `onMonthChange` is defined', () => {
      it('returns prop `activeMonth`', () => {
        props = { activeMonth: date, onMonthChange: jest.fn() }
        wrapper = shallow(<Calendar {...props} />)

        expect(wrapper.instance()._activeMonth()).toBe(props.activeMonth)
      })
    })

    describe('when prop `onMonthChange` is `undefined`', () => {
      it('returns state `activeMonth`', () => {
        wrapper = shallow(<Calendar activeMonth={date} />)
        wrapper.setState({ activeMonth: '2015-01-01' })

        expect(wrapper.instance()._activeMonth()).toBe('2015-01-01')
      })
    })
  })

  describe('#_highlight', () => {
    describe('when prop `highlighted` is `undefined`', () => {
      it('returns object', () => {
        wrapper = shallow(<Calendar />)

        expect(wrapper.instance()._highlight()).toEqual({
          end: null,
          start: null
        })
      })
    })

    describe('when prop `highlighted` is defined', () => {
      it('returns object', () => {
        props = {
          highlighted: {
            end: new Date(2015, 7, 1),
            start: date
          }
        }
        wrapper = shallow(<Calendar {...props} />)

        expect(wrapper.instance()._highlight()).toEqual(props.highlighted)
      })

      describe('when prop `highlighted.start` is invalid date', () => {
        it('returns object', () => {
          props = {
            highlighted: {
              end: new Date(2015, 7, 1),
              start: 'abc'
            }
          }
          wrapper = shallow(<Calendar {...props} />)

          expect(wrapper.instance()._highlight()).toEqual({
            end: null,
            start: null
          })
        })
      })

      describe('when prop `highlighted.end` is invalid date', () => {
        it('returns object', () => {
          props = {
            highlighted: {
              end: 'abc',
              start: date
            }
          }
          wrapper = shallow(<Calendar {...props} />)

          expect(wrapper.instance()._highlight()).toEqual({
            end: null,
            start: null
          })
        })
      })
    })
  })

  describe('#_selection', () => {
    let instance

    beforeEach(() => {
      wrapper = shallow(<Calendar />)
      instance = wrapper.instance()
    })

    describe('when #_selectionStart and #_selectionEnd returns valid date', () => {
      it('returns object', () => {
        const start = date
        const end = new Date(2015, 7, 1)

        instance._selectionStart = () => start
        instance._selectionEnd = () => end

        expect(instance._selection()).toEqual({ end, start })
      })
    })

    describe('when #_selectionStart and #_selectionEnd returns invalid date', () => {
      it('returns object', () => {
        instance._selectionStart = () => date
        instance._selectionEnd = () => 'abc'

        expect(instance._selection()).toEqual({ end: null, start: null })
      })
    })
  })

  describe('#_selectionDate', () => {
    describe("when prop `mode` is 'single'", () => {
      it('returns prop `selected', () => {
        props = { selected: date }
        wrapper = shallow(<Calendar {...props} />)

        expect(wrapper.instance()._selectionDate('end')).toBe(props.selected)
      })
    })

    describe("when prop `mode` is 'range'", () => {
      const selected = { end: new Date(2015, 7, 1), start: date }
      const selection = {
        end: new Date(2017, 7, 1),
        start: new Date(2017, 6, 1)
      }

      describe('when prop `onSelectionProgress` is `undefined`', () => {
        beforeEach(() => {
          props = { mode: 'range', selected }
          wrapper = shallow(<Calendar {...props} />)
          instance = wrapper.instance()
        })

        describe('when state `selection` is `defined`', () => {
          it('returns state `selection.start`', () => {
            wrapper.setState({ selection })

            expect(instance._selectionDate('start')).toBe(selection.start)
          })
        })

        it('returns prop `selected.start`', () => {
          expect(instance._selectionDate('start')).toBe(props.selected.start)
        })
      })

      describe('when prop `onSelectionProgress` is defined', () => {
        it('returns prop `selected.start`', () => {
          props = { mode: 'range', onSelectionProgress: () => { }, selected }
          wrapper = shallow(<Calendar {...props} />)
          wrapper.setState({ selection })

          expect(wrapper.instance()._selectionDate('start')).toBe(
            props.selected.start
          )
        })
      })
    })
  })

  describe('#_selectionChanged', () => {
    const selection = { end: new Date(2015, 7, 1), start: date }

    describe("when prop `mode` is 'single'", () => {
      describe('when prop `onSelect` is defined', () => {
        describe('when selection `start` is defined', () => {
          it('calls prop #onSelect', () => {
            props = { onSelect: jest.fn() }
            wrapper = shallow(<Calendar {...props} />)
            wrapper.instance()._selectionChanged(selection)

            expect(props.onSelect).toHaveBeenCalledTimes(1)

            expect(props.onSelect).toHaveBeenCalledWith(selection.start)
          })
        })
      })
    })

    describe("when prop `mode` is 'range'", () => {
      describe('when prop `onSelect` is defined', () => {
        it('calls prop #onSelect', () => {
          props = { mode: 'range', onSelect: jest.fn() }
          wrapper = shallow(<Calendar {...props} />)
          wrapper.instance()._selectionChanged(selection)

          expect(props.onSelect).toHaveBeenCalledTimes(1)

          expect(props.onSelect).toHaveBeenCalledWith(selection)
        })
      })

      describe('when prop `onSelectionProgress` is defined', () => {
        it('calls prop #onSelectionProgress', () => {
          props = { mode: 'range', onSelectionProgress: jest.fn() }
          wrapper = shallow(<Calendar {...props} />)
          wrapper
            .instance()
            ._selectionChanged(Object.assign(selection, { inProgress: true }))

          expect(props.onSelectionProgress).toHaveBeenCalledTimes(1)

          expect(props.onSelectionProgress).toHaveBeenCalledWith(
            Object.assign(selection, { inProgress: true })
          )
        })
      })

      describe('when prop `onSelectionProgress` is `undefined`', () => {
        beforeEach(() => {
          props = { mode: 'range' }
          wrapper = shallow(<Calendar {...props} />)
          instance = wrapper.instance()
        })

        describe('when selection `inProgress` is `true`', () => {
          it('sets state `selection`', () => {
            instance._selectionChanged(
              Object.assign(selection, { inProgress: true })
            )

            expect(wrapper.state('selection')).toEqual({
              end: new Date(2015, 7, 1),
              start: date
            })
          })
        })

        describe('when selection `inProgress` is `false`', () => {
          it('sets state `selection`', () => {
            instance._selectionChanged(
              Object.assign(selection, { inProgress: false })
            )

            expect(wrapper.state('selection')).toBe(null)
          })
        })
      })
    })
  })

  describe('#_noticeChanged', () => {
    it('sets state', () => {
      wrapper = shallow(<Calendar />)

      expect(wrapper.state('shownNoticeType')).toBe(null)
      wrapper.instance()._noticeChanged('disabled_day_click')

      expect(wrapper.state('shownNoticeType')).toBe('disabled_day_click')
    })
  })

  describe('#_today', () => {
    describe('when prop `today` is defined', () => {
      it('returns date', () => {
        wrapper = shallow(<Calendar today={date} />)

        expect(wrapper.instance()._today()).toBe(date)
      })
    })

    describe('when prop `today` is `undefined`', () => {
      it('returns date', () => {
        wrapper = shallow(<Calendar />)

        expect(wrapper.instance()._today()).toEqual(new Date(2000, 0, 1))
      })
    })
  })

  describe('#render', () => {
    it('renders <Calendar />', () => {
      const tree = ReactTestRenderer.create(
        <Calendar activeMonth={date} onMonthChange={() => { }} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  it('uses normalized minDate with a start at 00:00', () => {
    const dirtyMinDate = '2020-09-03T11:26:49.526Z'
    const { root } = ReactTestRenderer.create(
      <Calendar minDate={dirtyMinDate} activeMonth={date} onMonthChange={() => { }} />
    )

    const startOfMinDate = new Date(2020, 8, 3) // start of the day in local time - 00:00

    expect(root.findByType(Month).props.minDate).toEqual(startOfMinDate)
  })
})
