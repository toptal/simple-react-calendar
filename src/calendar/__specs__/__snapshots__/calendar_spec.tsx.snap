// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar #render renders <Calendar /> 1`] = `
<div
  className="calendar"
>
  <div
    className="calendar-month_header"
  >
    <button
      className="calendar-header_button is-prev"
      disabled={false}
      onClick={[Function]}
      title="Previous month"
      type="button"
    />
    <div
      className="calendar-month_header_title"
    >
      July 2015
    </div>
    <button
      className="calendar-header_button is-next"
      disabled={false}
      onClick={[Function]}
      title="Next month"
      type="button"
    />
  </div>
  <div
    className="calendar-month"
  >
    <div
      className="calendar-days_of_week"
    >
      <div
        className="calendar-days_of_week_day"
      >
        Mon
      </div>
      <div
        className="calendar-days_of_week_day"
      >
        Tue
      </div>
      <div
        className="calendar-days_of_week_day"
      >
        Wed
      </div>
      <div
        className="calendar-days_of_week_day"
      >
        Thu
      </div>
      <div
        className="calendar-days_of_week_day"
      >
        Fri
      </div>
      <div
        className="calendar-days_of_week_day is-weekend"
      >
        Sat
      </div>
      <div
        className="calendar-days_of_week_day is-weekend"
      >
        Sun
      </div>
    </div>
    <div
      className="calendar-week"
    >
      <button
        className="calendar-week-day is-prev_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-06-29"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-06-29"
      >
        29
      </button>
      <button
        className="calendar-week-day is-prev_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-06-30"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-06-30"
      >
        30
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-01"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-01"
      >
        1
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-02"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-02"
      >
        2
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-03"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-03"
      >
        3
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-weekend"
        data-simple-react-calendar-day="2015-07-04"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-04"
      >
        4
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-weekend"
        data-simple-react-calendar-day="2015-07-05"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-05"
      >
        5
      </button>
    </div>
    <div
      className="calendar-week"
    >
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-06"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-06"
      >
        6
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-07"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-07"
      >
        7
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-08"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-08"
      >
        8
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-09"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-09"
      >
        9
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-10"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-10"
      >
        10
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-weekend"
        data-simple-react-calendar-day="2015-07-11"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-11"
      >
        11
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-weekend"
        data-simple-react-calendar-day="2015-07-12"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-12"
      >
        12
      </button>
    </div>
    <div
      className="calendar-week"
    >
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-13"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-13"
      >
        13
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-14"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-14"
      >
        14
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-15"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-15"
      >
        15
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-16"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-16"
      >
        16
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-17"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-17"
      >
        17
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-weekend"
        data-simple-react-calendar-day="2015-07-18"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-18"
      >
        18
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-weekend"
        data-simple-react-calendar-day="2015-07-19"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-19"
      >
        19
      </button>
    </div>
    <div
      className="calendar-week"
    >
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-20"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-20"
      >
        20
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-21"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-21"
      >
        21
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-22"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-22"
      >
        22
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-23"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-23"
      >
        23
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-24"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-24"
      >
        24
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-weekend"
        data-simple-react-calendar-day="2015-07-25"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-25"
      >
        25
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-weekend"
        data-simple-react-calendar-day="2015-07-26"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-26"
      >
        26
      </button>
    </div>
    <div
      className="calendar-week"
    >
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-27"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-27"
      >
        27
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-28"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-28"
      >
        28
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-29"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-29"
      >
        29
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-30"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-30"
      >
        30
      </button>
      <button
        className="calendar-week-day is-current_month is-selectable is-working_day"
        data-simple-react-calendar-day="2015-07-31"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-07-31"
      >
        31
      </button>
      <button
        className="calendar-week-day is-next_month is-selectable is-weekend"
        data-simple-react-calendar-day="2015-08-01"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-08-01"
      >
        1
      </button>
      <button
        className="calendar-week-day is-next_month is-selectable is-weekend"
        data-simple-react-calendar-day="2015-08-02"
        onClick={[Function]}
        onMouseEnter={[Function]}
        type="button"
        value="2015-08-02"
      >
        2
      </button>
    </div>
  </div>
</div>
`;
